align/cell_fabric/gen_lef.py

```diff
@@ -2,138 +2,76 @@
 import json
 import collections
 
-def gen_lef_data(data, fp, macro_name, cell_pin, bodyswitch):
-    def s(x):
-        return "%.4f" % (x/10000.0)
+def lef_from_layout_d(layout_d, fp, out_lef, cell_pin, bodyswitch, blockM, *, exclude_layers, Scale_factor, m1pitch, m2pitch):
 
-    fp.write("MACRO %s\n" % macro_name)
-    fp.write("  ORIGIN 0 0 ;\n")
-    fp.write("  FOREIGN %s 0 0 ;\n" % macro_name)
+    bbox = layout_d['bbox']
 
-    fp.write("  SIZE %s BY %s ;\n" % (s(data['bbox'][2]), s(data['bbox'][3])))
+    assert (bbox[3]-bbox[1]) % m2pitch == 0, f"Cell height not a multiple of the grid {bbox}"
+    assert (bbox[2]-bbox[0]) % m1pitch == 0, f"Cell width not a multiple of the grid {bbox}"
 
-    exclude_layers = {"via0", "via1", "via2", "poly", "LISD", "SDT", "RVT",
-                      "M0", "fin", "polycon", "GCUT", "active", "nselect", "pselect", "nwell"}
+    fp.write("MACRO %s\n" % out_lef)
+    fp.write("  UNITS \n")
+    fp.write("    DATABASE MICRONS UNITS %s ;\n" % (1000*Scale_factor))
+    fp.write("  END UNITS \n")
+    fp.write("  ORIGIN 0 0 ;\n")
+    fp.write("  FOREIGN %s 0 0 ;\n" % out_lef)
 
-# O(npins * nsegments) algorithm. Could be O(npins + nsegments) FIX!
+    fp.write("  SIZE %s BY %s ;\n" % (bbox[2], bbox[3]))
+    cell_pin = list(cell_pin)
 
     for i in cell_pin:
         if i == 'B' and bodyswitch==0:continue
         fp.write("  PIN %s\n" % i)
-        #fp.write( "    DIRECTION %s ;\n" % obj['ported'])
+        #fp.write( "    DIRECTION %s ;\n" % obj[ported'])
         fp.write("    DIRECTION INOUT ;\n")
         fp.write("    USE SIGNAL ;\n")
-        fp.write("  PORT\n")
-        for obj in data['terminals']:
-            if 'pin' in obj:
-                if obj['pin'] == i:
-                    fp.write("      LAYER %s ;\n" % obj['layer'])
-                    fp.write("        RECT %s %s %s %s ;\n" %
-                             tuple([s(x) for x in obj['rect']]))
+        fp.write("    PORT\n")
+        for obj in layout_d['terminals']:
+            if 'pin' in obj and obj['pin'] == i:
+                fp.write("      LAYER %s ;\n" % obj['layer'])
+                fp.write("        RECT %s %s %s %s ;\n" % tuple(obj['rect']))
+                # Check Pins are on grid or not
+                if obj['layer'] == 'M2':
+                    cy = (obj['rect'][1]+obj['rect'][3])//2
+                    assert cy % m2pitch == 0, f"M2 pin is not on grid {cy} {cy%m2pitch}"
+                if obj['layer'] == 'M1' or obj['layer'] == 'M3':
+                    cx = (obj['rect'][0]+obj['rect'][2])//2
+                    assert cx % m1pitch == 0, "M1 pin is not on grid {cx} {cx%m1pitch}"
```

Comment:
Same comment applies here.

Reply:
So we should use the grid mechanism within cell fabric to check the grid. I'm all for that. I did this in a hacking way because it was already being done here and it was the particular problem that I was trying to find in the LEF file that Jitesh gave us.

I think we should have a general routine that takes a layout_d and a canvas and checks for off-grid errors. They we wouldn't need to embed this code in two different places in the PR.